### PR TITLE
Use crypto drivers for Bluetooth

### PR DESCRIPTION
**Pull request by lpereira**
In addition to making the Bluetooth subsystem work with crypto drivers, this PR makes a few changes to accommodate it:

- This adds all the primitives needed to support the Bluetooth subsystem, including new algorithms (PRNG, CMAC, and ECC-DH).
- Implements the new algorithms and their associated modes in both the TinyCrypt and mbedTLS shims
- Changes the Bluetooth subsystem to not require TinyCrypt, but rather declare its need for particular algorithms; the crypto backend can then be selected.  Declaring the need for a particular algorithm will enable/disable the features in either mbedTLS or TinyCrypt.  In the case of mbedTLS, however, only the default configuration file is aware of these declarations.

This has only been build-tested; I don't know enough of Bluetooth to carry on the necessary testing to ensure that this is working correctly, and I would appreciate help in getting this tested.

This is a step in closing #8081.

### CODE REVIEW

File: drivers/crypto/Kconfig [hunk 1]
```diff
@@ -128,4 +122,43 @@ endchoice
...
+if CRYPTO_SHIM
+
+config CRYPTO_ENABLE_AES
+	bool "Enable AES operations"
+	default y
+
+if CRYPTO_ENABLE_AES
+config CRYPTO_ENABLE_AES_CBC
+	bool "Enable CBC (block) mode for AES operations"
+	default n
```
Comment (ulfalizer) [inline drivers/crypto/Kconfig]: I know Zephyr has this all over, but just a small note: `bool` symbols implicitly default to `n`. `default n` is almost always redundant.

**Review by vakulgarg — CHANGES_REQUESTED** [review]
In commit log, rather than saying 10%, pls specify additional memory requirements in terms of bytes/KB. The percentage could vary depending upon how much total size we are assuming.

File: drivers/crypto/crypto_mtls_shim.c [hunk 2]
```diff
@@ -31,22 +31,36 @@
...
 #include <mbedtls/bignum.h>
 #include <mbedtls/ecdh.h>
+#endif
 
 #define MTLS_SUPPORT (CAP_RAW_KEY | CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS)
 
 struct mtls_shim_session {
 	enum cipher_algo algo;
 	union {
 		mbedtls_ccm_context mtls;
```
Comment (vakulgarg) [inline drivers/crypto/crypto_mtls_shim.c]: Since mtls is not covered under #ifdef like others, does it mean that aes-ccm is always enabled in the shim?

File: drivers/crypto/crypto_mtls_shim.c [hunk 3]
```diff
@@ -163,12 +180,19 @@ static void mtls_free_session(struct mtls_shim_session *session)
 	case CRYPTO_CIPHER_ALGO_AES:
 		mbedtls_ccm_free(&session->mtls);
 		break;
+
 	case CRYPTO_CIPHER_ALGO_PRNG:
+#if defined(CONFIG_MBEDTLS_ENABLE_HMAC_DRBG)
```
Comment (vakulgarg) [inline drivers/crypto/crypto_mtls_shim.c]: #ifdef should be before 'case' statement

File: subsys/bluetooth/host/hci_ecc.c [hunk 4]
```diff
@@ -166,13 +242,14 @@ static void emulate_le_generate_dhkey(void)
 	struct net_buf *buf;
 	int ret;
 
-	ret = uECC_valid_public_key(ecc.pk, &curve_secp256r1);
-	if (ret < 0) {
+	if (!is_public_key_valid(ecc.pk, sizeof(ecc.pk))) {
 		BT_ERR("public key is not valid (ret %d)", ret);
```
Comment (jhedberg) [inline subsys/bluetooth/host/hci_ecc.c]: You'll need to update this log since `ret` is undefined here.

**Comment by lpereira** [comment]
@vakulgarg I've added more detailed size measurements in one of the commit messages; I'm copying it here so it's easier to find:

>             Sample: mesh      Target: bbc_microbit
> 
>             Backend: tinycrypt
>             Memory region         Used Size  Region Size  %age Used
>                        FLASH:       98040 B       256 KB     37.40%
>                         SRAM:       15760 B        16 KB     96.19%
>                     IDT_LIST:         132 B         2 KB      6.45%
> 
>             Backend: mbedTLS
>             Memory region         Used Size  Region Size  %age Used
>                        FLASH:      124200 B       256 KB     47.38%
>                         SRAM:       15932 B        16 KB     97.24%
>                     IDT_LIST:         132 B         2 KB      6.45%
> 
>             Sample: mesh_demo     Target: bbc_microbit
> 
>             Backend: tinycrypt
>             Memory region         Used Size  Region Size  %age Used
>                        FLASH:       93904 B       256 KB     35.82%
>                         SRAM:       14548 B        16 KB     88.79%
>                     IDT_LIST:         132 B         2 KB      6.45%
> 
>             Backend: mbedTLS
>             Memory region         Used Size  Region Size  %age Used
>                        FLASH:      113004 B       256 KB     43.11%
>                         SRAM:       14720 B        16 KB     89.84%
>                     IDT_LIST:         132 B         2 KB      6.45%
>

**Comment by dleach02** [comment]
what are the plans for this?

**Comment by carlescufi** [comment]
@lpereira the size different is big enough that it will cause trouble on our chips, so this is a problem.

CC @sjanc @MariuszSkamra

**Comment by dleach02** [comment]
But the size difference between tiny and mbed where the application can choose which one it wants to use? Also, on the mbed, could this be that we are including stuff that isn't needed?

**Comment by carlescufi** [comment]
>But the size difference between tiny and mbed where the application can choose which one it wants to use? 

@dleach02 Not sure I understand this. I thought we'd replace TinyCrypt with mbedTLS, making the size increase essentially mandatory.

**Comment by dleach02** [comment]
The first comment above indicates that the changes here are to define a standard crypto front end interface and to hook that up to either tinycrypt or mbedtls. I'm all for the standard interface and need this PR for the work I'm doing on KW41Z

**Comment by carlescufi** [comment]
@dleach02 apologies, I misread the purpose of the PR, thanks for the explanation!

**Comment by dleach02** [comment]
I am curious though about the size differences. @lpereira, can you send me the two map files for this? (david.leach @ nxp.com)

**Comment by lpereira** [comment]
@dleach02 You can find the files [here](https://tia.mat.br/temp/bluetooth-map-files.tar.gz).  Warning: they decompress to the current directory.

**Comment by dleach02** [comment]
So the two things that get you when using mbedtls is the inclusion of bignum.c and des.c. These add 10K from .text and .rodata. The bignum.c states that it is required when doing ecc but I'm not sure why des.c is included?

Also, in general the mbedtls functions seem to be slightly bigger then tinycrypt.

**Comment by dleach02** [comment]
I agree with the PR objectives. I was just curious about what the sizing comparison is. I think we should have a separate PR to further break down mbedTLS based on the application needs as that would be good thing to do.

**Review by carlescufi — APPROVED** [review]
This is a really welcome change. I have no objections.

**Comment by carlescufi** [comment]
@lpereira can you rebase this?

**Comment by jhedberg** [comment]
I'd like to get PTS tests run on this before merging, to make sure we're not breaking anything with pairing or mesh provisioning.

**Comment by lpereira** [comment]
Rebased.

**Comment by carlescufi** [comment]
@vakulgarg mind taking another look?

**Comment by lpereira** [comment]
@jhedberg I appreciate this being tested on real devices.  I lack the knowledge to perform those myself.  I'll remove the DNM tag once this is confirmed working.

**Comment by MariuszSkamra** [comment]
@jhedberg @lpereira I wanted to test this yesterday but there were some compilation error, and I gave up. 
https://pastebin.com/rpvWJZfF

**Comment by lpereira** [comment]
@MariuszSkamra I've only test-built this for the bbc_microbit board; the mesh sample builds fine here.  Can you share which code you were building, board name, and Kconfig options different from the default for that board?

**Comment by MariuszSkamra** [comment]
@lpereira To test this against PTS (Bluetooth qualification tool) I need to have `tests/bluetooth/tester` application built. I'm using `nrf52840_pca10056`  board. Kconfig options are defined in `tests/bluetooth/tester/default.conf`

**Comment by lpereira** [comment]
@MariuszSkamra Thanks.  I've fixed the build and force-pushed.

**Review by d3zd3z — APPROVED** [review]
The crypto changes look good to me.

**Comment by d3zd3z** [comment]
As far as the size goes: I agree that maybe some changes could be made to the config file. In addition, it may be worth working with the upstream project if there are things in mbed TLS that could be made more configurable.

**Comment by lpereira** [comment]
Fixed some issues found in CI.  Rebased, and repushed.

File: drivers/crypto/crypto_mtls_shim.c [hunk 5]
```diff
@@ -189,17 +591,22 @@ static int mtls_query_caps(struct device *dev)
...
 static struct crypto_driver_api mtls_crypto_funcs = {
 	.begin_session = mtls_session_setup,
-	.free_session = mtls_session_free,
+	.free_session = mtls_session_dev_free,
 	.crypto_async_callback_set = NULL,
 	.query_hw_caps = mtls_query_caps,
 };
 
-DEVICE_AND_API_INIT(crypto_mtls, CONFIG_CRYPTO_MBEDTLS_SHIM_DRV_NAME,
+DEVICE_AND_API_INIT(crypto_mtls, CONFIG_CRYPTO_NAME,
```
Comment (ceolin) [inline drivers/crypto/crypto_mtls_shim.c]: Shouldn't this be changed in https://github.com/zephyrproject-rtos/zephyr/pull/8351/commits/13c522400e9ec80282d3a99bf27fab2821c41e99 ?

File: subsys/bluetooth/host/crypto.c [hunk 6]
```diff
@@ -45,14 +45,19 @@ static int prng_reseed(struct tc_hmac_prng_struct *h)
...
 
+		explicit_bzero(rp->rand, 8);
 		net_buf_unref(rsp);
 	}
 
-	extra = k_uptime_get();
+	ret = cipher_prng_op(&prng_session, &(struct cipher_prng_pkt) {
+		.reseed = true,
+		.additional_input = seed,
+		.additional_input_len = sizeof(seed),
```
Comment (ceolin) [inline subsys/bluetooth/host/crypto.c]: it will not work with tinycrypt.  tc_hmac_prng_reseed() will require .data anda .data_len set.
additional_input should be mapped to extra and not to seed, the problem is that medtls is expecting seed be passed in additional_input. This must be changed too.

File: subsys/bluetooth/host/crypto.c [hunk 7]
```diff
@@ -78,83 +83,127 @@ int prng_init(void)
...
 int bt_encrypt_le(const u8_t key[16], const u8_t plaintext[16],
 		  u8_t enc_data[16])
 {
-	struct tc_aes_key_sched_struct s;
-	u8_t tmp[16];
+	u8_t le_key[16];
+	u8_t le_plaintext[16];
+	struct cipher_ctx ctx = {
+		.key.bit_stream = le_key,
+		.keylen = 128 / 8,
```
Comment (ceolin) [inline subsys/bluetooth/host/crypto.c]: why aren't you using `16` just like you are doing below ?

File: drivers/crypto/crypto_mtls_shim.c [hunk 8]
```diff
@@ -99,37 +142,48 @@ static int mtls_ccm_decrypt_auth(struct cipher_ctx *ctx,
...
+	case CRYPTO_CIPHER_ALGO_ECC:
+		/* Do nothing. */
+		break;
 	}
 
-	if (algo != CRYPTO_CIPHER_ALGO_AES) {
-		SYS_LOG_ERR("Unsupported algo");
-		return -EINVAL;
-	}
+	memset(session, 0, sizeof(*session));
```
Comment (ceolin) [inline drivers/crypto/crypto_mtls_shim.c]: isn't the case of using `explicit_bzero()` ?

File: drivers/crypto/crypto_mtls_shim.c [hunk 9]
```diff
@@ -171,13 +222,364 @@ static int mtls_session_setup(struct device *dev, struct cipher_ctx *ctx,
...
+	mbedtls_ecp_group_free(&grp);
+
+	return r == 0 ? 0 : -EINVAL;
+}
+
+static int mtls_ecc_gen_shared_key(struct cipher_ctx *ctx,
+				   struct cipher_ecc_pkt *pkt)
+{
+	mbedtls_ecp_keypair keypair;
+	mbedtls_ecp_point Q;
```
Comment (ceolin) [inline drivers/crypto/crypto_mtls_shim.c]: question, why this variable is uppercase ?

File: drivers/crypto/crypto_mtls_shim.c [hunk 10]
```diff
@@ -171,13 +222,364 @@ static int mtls_session_setup(struct device *dev, struct cipher_ctx *ctx,
...
+			      struct cipher_ctx *ctx,
+			      enum cipher_algo algo,
+			      enum cipher_mode mode,
+			      enum cipher_op op_type)
+{
+	ARG_UNUSED(dev);
+
+	if (ctx->flags & ~(MTLS_SUPPORT)) {
+		SYS_LOG_ERR("Unsupported flag");
+		return -EINVAL;
```
Comment (ceolin) [inline drivers/crypto/crypto_mtls_shim.c]: Isn't the case of  `-ENOTSUP` ? It seems exactly the same thing of an invalid algorithm (line 572)

File: include/crypto/cipher_structs.h [hunk 11]
```diff
@@ -103,6 +124,11 @@ struct cipher_ctx {
 		 * available to caller
 		 */
 		void *handle;
+		/* For PRNG operations */
+		struct {
+			u8_t *data;
+			size_t len;
+		} personalization;
```
Comment (ceolin) [inline include/crypto/cipher_structs.h]: I saw around the repo, two fields, `personalization_data` and `personalization_data_len`. Wouldn't be better keep this pattern ?

File: subsys/bluetooth/host/crypto.c [hunk 12]
```diff
@@ -78,83 +83,127 @@ int prng_init(void)
...
+	};
+	struct cipher_pkt pkt = {
+		.in_buf = (const u8_t *)le_plaintext,
```
Comment (ceolin) [inline subsys/bluetooth/host/crypto.c]: `.in_buf` is not a const

File: subsys/bluetooth/host/crypto.c [hunk 13]
```diff
@@ -78,83 +83,127 @@ int prng_init(void)
...
-	return 0;
+	return ret;
 }
 
 int bt_encrypt_be(const u8_t key[16], const u8_t plaintext[16],
 		  u8_t enc_data[16])
 {
-	struct tc_aes_key_sched_struct s;
+	struct cipher_ctx ctx = {
+		.key.bit_stream = key,
```
Comment (ceolin) [inline subsys/bluetooth/host/crypto.c]: warning: initialization discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]

File: subsys/bluetooth/host/CMakeLists.txt [hunk 14]
```diff
@@ -1,14 +1,14 @@
...
 
-zephyr_library_sources_ifdef(CONFIG_BT_HCI_RAW          hci_raw.c)
-zephyr_library_sources_ifdef(CONFIG_BT_DEBUG_MONITOR    monitor.c)
-zephyr_library_sources_ifdef(CONFIG_BT_TINYCRYPT_ECC    hci_ecc.c)
-zephyr_library_sources_ifdef(CONFIG_BT_A2DP             a2dp.c)
-zephyr_library_sources_ifdef(CONFIG_BT_AVDTP            avdtp.c)
-zephyr_library_sources_ifdef(CONFIG_BT_RFCOMM           rfcomm.c)
-zephyr_library_sources_ifdef(CONFIG_BT_TESTING          testing.c)
-zephyr_library_sources_ifdef(CONFIG_BT_SETTINGS         settings.c)
+zephyr_library_sources_ifdef(CONFIG_BT_HCI_RAW           hci_raw.c)
```
Comment (MariuszSkamra) [inline subsys/bluetooth/host/CMakeLists.txt]: those cosmetic changes are not part of this PR I guess

**Comment by lpereira** [comment]
Excellent, @MariuszSkamra, thanks for testing!  Results are better than I expected, to be honest.  I don't know much about Bluetooth, so short of addressing the assertions, I would need help to get every PTS test to pass again.

File: drivers/crypto/crypto_mtls_shim.c (hunk 9 above)
Comment (lpereira) [reply drivers/crypto/crypto_mtls_shim.c]: This is what mbedtls uses; I used this for consistency.  That's also the same notation used in ECDH math.  I guess I could change this to `public_key`, but then it would be inconsistent with `struct mbedtls_ecp_keypair` and other related types

File: drivers/crypto/crypto_mtls_shim.c (hunk 8 above)
Comment (lpereira) [reply drivers/crypto/crypto_mtls_shim.c]: Doesn't hurt to use it here, yes.

File: drivers/crypto/crypto_mtls_shim.c (hunk 9 above)
Comment (ceolin) [reply drivers/crypto/crypto_mtls_shim.c]: :+1:

File: drivers/crypto/crypto_tc_shim.c [hunk 15]
```diff
@@ -8,22 +8,36 @@
...
+};
 
-#define CRYPTO_MAX_SESSION CONFIG_CRYPTO_TINYCRYPT_SHIM_MAX_SESSION
-
-static struct tc_shim_drv_state tc_driver_state[CRYPTO_MAX_SESSION];
+K_MEM_POOL_DEFINE(tc_shim_session_pool,
+		  sizeof(struct tc_shim_drv_state),
+		  sizeof(struct tc_shim_drv_state),
+		  CONFIG_CRYPTO_MAX_SESSION,
+		  __alignof__(struct tc_shim_drv_state));
```
Comment (ceolin) [inline drivers/crypto/crypto_tc_shim.c]: same thing mentioned in mbedtls_shim.

**Comment by ceolin** [comment]
@lpereira as just point the things I'll fix in this pr. I'll rebase it apply the fixes I mentioned here.

**Comment by lpereira** [comment]
I've removed the DNM as this seems to be a generally accepted approach.  @ceolin is ironing out some issues in this patch and will send a newer version soon; I'll close this PR once that happens.

**Comment by jhedberg** [comment]
Just a heads-up: #9181 will likely cause conflicts and you'll want to update this PR to follow suit (or possibly the other way around - depending on which PR gets merged first)

**Comment by ceolin** [comment]
@lpereira #9227

**Comment by lpereira** [comment]
Closing in favor of #9227 (@ceolin is taking over the series from me)
